Add event topic and fix sub manager docs

fselmo opened the pull request:
### What was wrong?

- The code block was not correct for subscription manager docs.
- Getting the contract event's topic was much too difficult.

### How was it fixed?

- Fix subscription manager docs.
- Add `topic` property to `BaseContractEvent`.
- Add topic tests.

### Todo:

- [x] Clean up commit history
- [x] Add or update documentation related to these changes
- [x] Add entry to the [release notes](https://github.com/ethereum/web3.py/blob/main/newsfragments/README.md)

#### Cute Animal Picture

<img width="724" alt="Screenshot 2025-01-15 at 11 40 56" src="https://github.com/user-attachments/assets/10b3de8d-6def-4540-83c7-4a6a697d5321" />

fselmo committed:
Add support for retrieving a contract event topic:

- Add ``topic`` property to ``ContractEvent`` to retrieve the topic
  of the event as ``encode_hex(keccak(text=event.signature))``.

fselmo committed:
Update current documentation for subscription manager.

fselmo committed:
Use event.topic in logs integration tests

kclowes reviewed (approved):
Sick! LGTM 🚀 

Edit: LGTM once you figure out the failing test :)

fselmo committed:
Only calculate the topic when it is asked for, while caching the result.

- Calculating the topic at ``__init__`` is not necessary. Instead,
  calculate it when the topic is asked for and set the internal
  ``_topic`` attribute only on the first call.

fselmo commented:
> Edit: LGTM once you figure out the failing test :)

It looked related to multiple contract init time test. Glad we have that in there. In order to not add to init times, I opted for calculating the topic as an `@property` and setting it only when it is called ([latest commit](https://github.com/ethereum/web3.py/pull/3578/commits/6afe5e1aa474f2f7fbfaaf6c380fe2c9a5438eb0))

reedsa reviewed (approved):
Great updates, lgtm!
